io/EptReader.cpp

```diff
@@ -0,0 +1,461 @@
+        });
+    }
+    else
+    {
+        for (uint64_t dir(0); dir < 8; ++dir)
+        {
+            overlaps(hier, key.bisect(dir));
+        }
+    }
+}
+
+PointViewSet EptReader::run(PointViewPtr view)
+{
+    uint64_t i(0);
+    for (const Key& key : m_overlapKeys)
+    {
+        log()->get(LogLevel::Debug) << "Data " << ++i << "/" <<
+            m_overlapKeys.size() << ": " << key.toString() << std::endl;
+
+        m_pool->add([this, &view, &key]()
+        {
+            if (m_info->dataType() == EptInfo::DataType::Laszip)
+                readLaszip(*view, key);
+            else
+                readBinary(*view, key);
+        });
+    }
+
+    m_pool->await();
+
+    log()->get(LogLevel::Debug) << "Done reading!" << std::endl;
+
+    PointViewSet views;
+    views.insert(view);
+    return views;
+}
+
+void EptReader::readLaszip(PointView& dst, const Key& key) const
+{
+    auto handle(m_ep->getLocalHandle(key.toString() + ".laz"));
+
+    PointTable table;
+
+    Options options;
+    options.add("filename", handle->localPath());
+    options.add("use_eb_vlr", true);
+
+    LasReader reader;
+    reader.setOptions(options);
+
+    std::lock_guard<std::mutex> lock(m_mutex);
+    reader.prepare(table);
+
+    // While the laszip data type will contain a scale/offset in the Entwine
+    // metadata, the LasReader will handle the transformation.
+    for (auto& src : reader.execute(table))
+    {
+        PointRef pr(*src);
+        for (uint64_t i(0); i < src->size(); ++i)
+        {
+            pr.setPointId(i);
+            process(dst, pr);
+        }
+    }
+}
+
+void EptReader::readBinary(PointView& dst, const Key& key) const
+{
+    auto data(m_ep->getBinary(key.toString() + ".bin"));
+    ShallowPointTable table(*m_remoteLayout, data.data(), data.size());
+    PointRef pr(table);
+
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    // Binary data scale/offset needs to be handled here.
+    for (uint64_t i(0); i < table.numPoints(); ++i)
+    {
+        pr.setPointId(i);
+        process(dst, pr, true);
+    }
+}
+
+void EptReader::process(PointView& dst, PointRef& pr, bool unscale) const
+{
+    const auto pointId(dst.size());
+    double x = pr.getFieldAs<double>(Dimension::Id::X);
+    double y = pr.getFieldAs<double>(Dimension::Id::Y);
+    double z = pr.getFieldAs<double>(Dimension::Id::Z);
+    const uint64_t o = pr.getFieldAs<uint64_t>(Dimension::Id::OriginId);
+
+    if (unscale)
+    {
+        x = x * m_info->scale()[0] + m_info->offset()[0];
+        y = y * m_info->scale()[1] + m_info->offset()[1];
+        z = z * m_info->scale()[2] + m_info->offset()[2];
+    }
+
+    const bool selected = !m_queryOriginId || o == *m_queryOriginId;
+
+    if (selected && m_queryBounds.contains(x, y, z))
+    {
+        dst.setField(Dimension::Id::X, pointId, x);
+        dst.setField(Dimension::Id::Y, pointId, y);
+        dst.setField(Dimension::Id::Z, pointId, z);
+
+        char* pos(dst.getPoint(pointId));
```

Comment:
Are you referring to the usage of `getPoint` because we know we've just created that PointId above?  Would you prefer:

```
    if (selected && m_queryBounds.contains(x, y, z))
    {
        char* pos(dst.getOrAddPoint(pointId));

        dst.setField(Dimension::Id::X, pointId, x);
        dst.setField(Dimension::Id::Y, pointId, y);
        dst.setField(Dimension::Id::Z, pointId, z);

        ...
```

Reply:
No.  Just the '*' in the wrong place.  There's no problem here.